src/mscorlib/src/System/String.cs

```diff
@@ -159,27 +159,26 @@ public sealed class String : IComparable, ICloneable, IConvertible, IEnumerable
             Contract.Ensures(Contract.Result<String>() != null);
             Contract.EndContractBlock();
 
-            if (separator == null)
-                separator = String.Empty;
-
-
             using(IEnumerator<String> en = values.GetEnumerator()) {
                 if (!en.MoveNext())
                     return String.Empty;
 
-                StringBuilder result = StringBuilderCache.Acquire();
-                if (en.Current != null) {
-                    result.Append(en.Current);
-                }
+                String value = en.Current;
 
-                while (en.MoveNext()) {
-                    result.Append(separator);
-                    if (en.Current != null) {
+                if (en.MoveNext()) { 
```

Comment:
We could probably get rid of the 'value' field and reduce the code indentation something like this

```csharp
if (!en.MoveNext()) {
  return en.Current == null? String.Empty : en.Current;
}

StringBuilder result = StringBuilderCache.Acquire();
result.Append(en.Current);
do {
    result.Append(separator);
    result.Append(en.Current);
} while (en.MoveNext());
return StringBuilderCache.GetStringAndRelease(result);
```

I know we are calling en.Current twice in the first if, and if we think that is causing an issue we can capture it in a local field.
What do you think?